File: engine/src/Scripting/Behaviour.cpp
#include "Behaviour.h"
#include "Debug/Logger.h"

namespace Engine {
    Behaviour::Behaviour(sol::state& lua_state, const std::string& script_path) {
        sol::load_result script = lua_state.load_file(script_path);
        if (!script.valid()) {
            sol::error err = script;
            ERROR(err.what());
            return;
        }

        self = script();

        // Setup Lifecycle Hooks
        hooks.start = self["start"];
        hooks.update = self["update"];
        hooks.late_update = self["late_update"];
        hooks.on_destroy = self["on_destroy"];
        hooks.on_enable = self["on_enable"];
        hooks.on_disable = self["on_disable"];

        // Setup Physics Hooks
        hooks.fixed_update = self["fixed_update"];
        hooks.on_trigger_enter = self["on_trigger_enter"];
        hooks.on_trigger_stay = self["on_trigger_stay"];
        hooks.on_trigger_exit = self["on_trigger_exit"];
        hooks.on_collision_enter = self["on_collision_enter"];
        hooks.on_collision_stay = self["on_collision_stay"];
        hooks.on_collision_exit = self["on_collision_exit"];

        // Setup Rendering Hooks
        hooks.on_draw_gizmos = self["on_draw_gizmos"];
        hooks.on_gui = self["on_gui"];
    }

    Behaviour::~Behaviour() {
        on_destroy();
    }

    // Lifecycle Hooks
    void Behaviour::start() {
        if (hooks.start.valid()) {
            hooks.start(self);
        }
    }

    void Behaviour::update(float delta) {
        if (hooks.update.valid()) {
            hooks.update(self, delta);
        }
    }

    void Behaviour::late_update(float delta) {
        if (hooks.late_update.valid()) {
            hooks.late_update(self, delta);
        }
    }

    void Behaviour::on_destroy() {
        if (hooks.on_destroy.valid()) {
            hooks.on_destroy(self);
        }
    }

    void Behaviour::on_enable() {
        if (hooks.on_enable.valid()) {
            hooks.on_enable(self);
        }
    }

    void Behaviour::on_disable() {
        if (hooks.on_disable.valid()) {
            hooks.on_disable(self);
        }
    }

    // Physics Hooks
    void Behaviour::fixed_update(float delta) {
        if (hooks.fixed_update.valid()) {
            hooks.fixed_update(self, delta);
        }
    }

    void Behaviour::on_trigger_enter() {
        if (hooks.on_trigger_enter.valid()) {
            hooks.on_trigger_enter(self);
        }
    }

    void Behaviour::on_trigger_stay() {
        if (hooks.on_trigger_stay.valid()) {
            hooks.on_trigger_stay(self);
        }
    }

    void Behaviour::on_trigger_exit() {
        if (hooks.on_trigger_exit.valid()) {
            hooks.on_trigger_exit(self);
        }
    }

    void Behaviour::on_collision_enter() {
        if (hooks.on_collision_enter.valid()) {
            hooks.on_collision_enter(self);
        }
    }

    void Behaviour::on_collision_stay() {
        if (hooks.on_collision_stay.valid()) {
            hooks.on_collision_stay(self);
        }
    }

    void Behaviour::on_collision_exit() {
        if (hooks.on_collision_exit.valid()) {
            hooks.on_collision_exit(self);
        }
    }

    // Rendering Hooks
    void Behaviour::on_draw_gizmos() {
        if (hooks.on_draw_gizmos.valid()) {
            hooks.on_draw_gizmos(self);
        }
    }

    void Behaviour::on_gui() {
        if (hooks.on_gui.valid()) {
            hooks.on_gui(self);
        }
    }
}

File: tests/project/library/behaviour.lua
---@class Behaviour
---@field class string
Behaviour = {
    class = "Behaviour"
}

-- Constructor

function Behaviour:new ()
    local obj = {}
    setmetatable(obj, self)
    self.__index = self
    return obj
end

-- Lifecycle Hooks

---Runs once when the script starts up
---@return nil
function Behaviour:start()
end

---Runs at the start of every frame
---@param delta number?
---@return nil
function Behaviour:update(delta)
end

---Runs at the end of every frame
---@param delta number?
---@return nil
function Behaviour:late_update(delta)
end

---Runs when the component or entity is destroyed
---@return nil
function Behaviour:on_destroy()
end

---Runs when the component is enabled
---@return nil
function Behaviour:on_enable()
end

---Runs when the coponent is disabled
---@return nil
function Behaviour:on_disable()
end

-- Physics Hooks

---Runs every physics frame
---@param delta number?
---@return nil
function Behaviour:fixed_update(delta)
end

---Runs when entering a trigger
---@return nil
function Behaviour:on_trigger_enter()
end

---Runs when on a trigger for more than one frame
---@return nil
function Behaviour:on_trigger_stay()
end

---Runs when exiting a trigger
---@return nil
function Behaviour:on_trigger_exit()
end

---Runs when entering a collision
---@return nil
function Behaviour:on_collision_enter()
end

---Runs when on a collision for more than one frame
---@return nil
function Behaviour:on_collision_stay()
end

---Runs when exiting a collision
---@return nil
function Behaviour:on_collision_exit()
end

-- Rendering hooks

---Runs when drawing guizmos
---@return nil
function Behaviour:on_draw_gizmos()
end

---Runs when drawing Gui
---@return nil
function Behaviour:on_gui()
end

Behaviour.__index = Behaviour